tiledb/sm/filesystem/vfs.cc

```diff
@@ -811,66 +812,106 @@ Status VFS::terminate() {
 }
 
 Status VFS::ls(const URI& parent, std::vector<URI>* uris) const {
-  if (!init_)
-    return LOG_STATUS(Status_VFSError("Cannot list; VFS not initialized"));
+  auto&& [st, entries] = ls_with_sizes(parent);
+  RETURN_NOT_OK(st);
+
+  for (auto& fs : *entries) {
+    uris->emplace_back(fs.path().to_string());
+  }
+
+  return Status::Ok();
+}
+
+tuple<Status, optional<std::vector<FileStat>>> VFS::ls_with_sizes(
+    const URI& parent) const {
+  if (!init_) {
+    auto st = LOG_STATUS(Status_VFSError("Cannot list; VFS not initialized"));
+    return {st, std::nullopt};
+  }
 
   // Noop if `parent` is not a directory, do not error out.
   // For S3, GCS and Azure, `ls` on a non-directory will just
   // return an empty `uris` vector.
   if (!(parent.is_s3() || parent.is_gcs() || parent.is_azure())) {
     bool flag = false;
-    RETURN_NOT_OK(is_dir(parent, &flag));
-    if (!flag)
-      return Status::Ok();
+    RETURN_NOT_OK_TUPLE(is_dir(parent, &flag), nullopt);
+
+    if (!flag) {
+      return {Status::Ok(), std::vector<FileStat>()};
+    }
   }
 
-  std::vector<std::string> paths;
+  optional<std::vector<FileStat>> entries;
   if (parent.is_file()) {
 #ifdef _WIN32
-    RETURN_NOT_OK(win_.ls(parent.to_path(), &paths));
-#else
-    RETURN_NOT_OK(posix_.ls(parent.to_path(), &paths));
-#endif
-  } else if (parent.is_hdfs()) {
-#ifdef HAVE_HDFS
-    RETURN_NOT_OK(hdfs_->ls(parent, &paths));
+    Status st;
+    std::tie(st, entries) = win_.ls_with_sizes(parent);
 #else
-    return LOG_STATUS(Status_VFSError("TileDB was built without HDFS support"));
+    Status st;
+    std::tie(st, entries) = posix_.ls_with_sizes(parent);
 #endif
+    if (!st.ok()) {
```

Comment:
Can be condensed to `RETURN_NOT_OK_TUPLE(st, nullopt)`

Reply:
Ups, missed these ones, will fix.